File: 2020-2021/Autumn/ProducersAndConsumers/src/com/company/store/Consumer.java
package com.company.store;

public class Consumer<E> extends Thread {
    private final Store<E> store;
    private Boolean running = false;

    public Consumer(Store<E> s) {
        store = s;
    }

    public void run() {
        running = true;
        while (running) {
            boolean fl = true;
            while (fl && running) {
                fl = store.get();
                try {
                    sleep(10);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
            try {
                sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public void stopConsumer() {
        running = false;
    }
}

File: Tasks/ThirdTerm/ProducerConsumer/ProducerConsumer.Library/Consumer.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;

namespace ProducerConsumer.Library
{
    public class Consumer<T> : IDisposable where T : class
    {
        private string name;
        private Thread thread;
        private Manager<T> manager;
        private volatile bool isWorking;

        public Consumer(Manager<T> manager, string name)
        {
            this.name = name;
            this.manager = manager;
            isWorking = true;
            thread = new Thread(Run);
        }

        private void Run()
        {
            while (isWorking)
            {
                manager.Pop(name);
                Thread.Sleep(1000);
            }
        }

        public void Start()
        {
            thread.Start();
        }

        public bool IsWorking()
        {
            return isWorking;
        }

        public void Exit()
        {
            Dispose();
        }

        public void Dispose()
        {
            isWorking = false;
            thread.Join();
        }
    }
}
